File: tests/basics/weakref_ref_collect.py
# Test weakref.ref() functionality requiring gc.collect().
# Should be kept in sync with tests/ports/webassembly/weakref_ref_collect.py.

try:
    import weakref
except ImportError:
    print("SKIP")
    raise SystemExit

# gc module must be available if weakref is.
import gc

# Cannot reference non-heap objects.
for value in (None, False, True, Ellipsis, 0, "", ()):
    try:
        weakref.ref(value)
    except TypeError:
        print(value, "TypeError")


class A:
    def __str__(self):
        return "<A object>"


def callback(r):
    print("callback", r())


def test():
    print("test basic use of ref() with only one argument")
    a = A()
    r = weakref.ref(a)
    print(r())
    a = None
    clean_the_stack = [0, 0, 0, 0]
    gc.collect()
    print(r())

    print("test use of ref() with a callback")
    a = A()
    r = weakref.ref(a, callback)
    print(r())
    a = None
    clean_the_stack = [0, 0, 0, 0]
    gc.collect()
    print(r())

    print("test when weakref gets collected before the object it refs")
    a = A()
    r = weakref.ref(a, callback)
    print(r())
    r = None
    clean_the_stack = [0, 0, 0, 0]
    gc.collect()
    a = None

    print("test a double reference")
    a = A()
    r1 = weakref.ref(a, callback)
    r2 = weakref.ref(a, callback)
    print(r1(), r2())
    a = None
    clean_the_stack = [0, 0, 0, 0]
    gc.collect()
    print(r1(), r2())


test()

File: tests/ports/webassembly/weakref_ref_collect.mjs
// Test weakref.ref() functionality requiring gc.collect().
// Should be kept in sync with tests/basics/weakref_ref_collect.py.
//
// This needs custom testing on the webassembly port since the GC can only
// run when Python code returns to JavaScript.

const mp = await (await import(process.argv[2])).loadMicroPython();

// Set up.
mp.runPython(`
import gc, weakref

class A:
    def __str__(self):
        return "<A object>"

def callback(r):
    print("callback", r())
`);

console.log("test basic use of ref() with only one argument");
mp.runPython(`
    a = A()
    r = weakref.ref(a)
    print(r())
    gc.collect()
`);
console.log("(outside Python)");
mp.runPython(`
    print(r())
    a = None
    gc.collect()
`);
console.log("(outside Python)");
mp.runPython(`
    print(r())
`);

console.log("test use of ref() with a callback");
mp.runPython(`
    a = A()
    r = weakref.ref(a, callback)
    print(r())
    gc.collect()
`);
console.log("(outside Python)");
mp.runPython(`
    print(r())
    a = None
    gc.collect()
`);
console.log("(outside Python)");
mp.runPython(`
    print(r())
`);

console.log("test when weakref gets collected before the object it refs");
mp.runPython(`
    a = A()
    r = weakref.ref(a, callback)
    print(r())
    r = None
    gc.collect()
`);
console.log("(outside Python)");
mp.runPython(`
    a = None
    gc.collect()
`);
